src/wasm-traversal.h

```diff
@@ -157,185 +149,209 @@ struct WasmWalkerBase : public WasmVisitor<SubType, ReturnType> {
+  static void doVisitHost(SubType* self, Expression** currp)         { self->visitExpression(*currp); self->visitHost((*currp)->cast<Host>()); }
+  static void doVisitNop(SubType* self, Expression** currp)          { self->visitExpression(*currp); self->visitNop((*currp)->cast<Nop>()); }
+  static void doVisitUnreachable(SubType* self, Expression** currp)  { self->visitExpression(*currp); self->visitUnreachable((*currp)->cast<Unreachable>()); }
 };
 
-//
-// Simple WebAssembly children-first walking (i.e., post-order, if you look
-// at the children as subtrees of the current node), with the ability to replace
-// the current expression node. Useful for writing optimization passes.
-//
+// Walks in post-order, i.e., children first. When there isn't an obvious
+// order to operands, we follow them in order of execution.
 
-template<typename SubType, typename ReturnType = void>
-struct WasmWalker : public WasmReplacerWalker<SubType, ReturnType> {
-  // By default, do nothing
-  ReturnType visitBlock(Block *curr) {}
-  ReturnType visitIf(If *curr) {}
-  ReturnType visitLoop(Loop *curr) {}
-  ReturnType visitBreak(Break *curr) {}
-  ReturnType visitSwitch(Switch *curr) {}
-  ReturnType visitCall(Call *curr) {}
-  ReturnType visitCallImport(CallImport *curr) {}
-  ReturnType visitCallIndirect(CallIndirect *curr) {}
-  ReturnType visitGetLocal(GetLocal *curr) {}
-  ReturnType visitSetLocal(SetLocal *curr) {}
-  ReturnType visitLoad(Load *curr) {}
-  ReturnType visitStore(Store *curr) {}
-  ReturnType visitConst(Const *curr) {}
-  ReturnType visitUnary(Unary *curr) {}
-  ReturnType visitBinary(Binary *curr) {}
-  ReturnType visitSelect(Select *curr) {}
-  ReturnType visitReturn(Return *curr) {}
-  ReturnType visitHost(Host *curr) {}
-  ReturnType visitNop(Nop *curr) {}
-  ReturnType visitUnreachable(Unreachable *curr) {}
+template<typename SubType>
+struct PostWalker : public Walker<SubType> {
 
-  ReturnType visitFunctionType(FunctionType *curr) {}
-  ReturnType visitImport(Import *curr) {}
-  ReturnType visitExport(Export *curr) {}
-  ReturnType visitFunction(Function *curr) {}
-  ReturnType visitTable(Table *curr) {}
-  ReturnType visitMemory(Memory *curr) {}
-  ReturnType visitModule(Module *curr) {}
+  static void scan(SubType* self, Expression** currp) {
 
-  // children-first
-  void walk(Expression*& curr) override {
-    if (!curr) return;
-
-    // special-case Block, because Block nesting (in their first element) can be incredibly deep
-    if (curr->is<Block>()) {
-      auto* block = curr->dyn_cast<Block>();
-      std::vector<Block*> stack;
-      stack.push_back(block);
-      while (block->list.size() > 0 && block->list[0]->is<Block>()) {
-        block = block->list[0]->cast<Block>();
-        stack.push_back(block);
+    Expression* curr = *currp;
+    switch (curr->_id) {
+      case Expression::Id::InvalidId: abort();
```

Comment:
If it's like `llvm_unreachable` which has basically the behavior you describe (i.e. it's compiled away in non-assert builds, and the optimizer assumes it's unreachable), then that sounds about right. AFAIK most of this codebase doesn't really have that distinction. Although it might be a useful one to start drawing, especially if you fold your assert/abort messages into unreachable as llvm's assert and unreachable do; then the release binary gets even better size benefits, which could be a big deal if we expect to be running this under wasm.